parquet/src/main/java/org/apache/iceberg/parquet/ParquetFormatModel.java

```diff
@@ -0,0 +1,392 @@
+import org.apache.iceberg.data.parquet.GenericParquetWriter;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.formats.FormatModel;
+import org.apache.iceberg.formats.ReadBuilder;
+import org.apache.iceberg.formats.WriteBuilder;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.DeleteSchemaUtil;
+import org.apache.iceberg.io.FileAppender;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.mapping.NameMapping;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.parquet.column.ParquetProperties;
+import org.apache.parquet.schema.MessageType;
+
+public class ParquetFormatModel<D, S, F> implements FormatModel<D, S> {
+  public static final String WRITER_VERSION_KEY = "parquet.writer.version";
+
+  private final Class<D> type;
+  private final Class<S> schemaType;
+  private final ReaderFunction<D> readerFunction;
+  private final BatchReaderFunction<D, F> batchReaderFunction;
+  private final WriterFunction<S> writerFunction;
+
+  private ParquetFormatModel(
+      Class<D> type,
+      Class<S> schemaType,
+      ReaderFunction<D> readerFunction,
+      BatchReaderFunction<D, F> batchReaderFunction,
+      WriterFunction<S> writerFunction) {
+    this.type = type;
+    this.schemaType = schemaType;
+    this.readerFunction = readerFunction;
+    this.batchReaderFunction = batchReaderFunction;
+    this.writerFunction = writerFunction;
+  }
+
+  public ParquetFormatModel(Class<D> type) {
+    this(type, null, null, null);
+  }
+
+  public ParquetFormatModel(
+      Class<D> type,
+      Class<S> schemaType,
+      ReaderFunction<D> readerFunction,
+      WriterFunction<S> writerFunction) {
+    this(type, schemaType, readerFunction, null, writerFunction);
+  }
+
+  public ParquetFormatModel(
+      Class<D> type, Class<S> schemaType, BatchReaderFunction<D, F> batchReaderFunction) {
+    this(type, schemaType, null, batchReaderFunction, null);
+  }
+
+  @Override
+  public FileFormat format() {
+    return FileFormat.PARQUET;
+  }
+
+  @Override
+  public Class<D> type() {
+    return type;
+  }
+
+  @Override
+  public Class<S> schemaType() {
+    return schemaType;
+  }
+
+  @Override
+  public WriteBuilder<D, S> writeBuilder(OutputFile outputFile) {
+    return new WriteBuilderWrapper<>(outputFile, writerFunction);
+  }
+
+  @Override
+  public ReadBuilder<D, S> readBuilder(InputFile inputFile) {
+    if (readerFunction != null) {
+      return new NonBatchReaderWrapper<>(inputFile, readerFunction);
+    } else if (batchReaderFunction != null) {
+      return new BatchReaderWrapper<>(inputFile, batchReaderFunction);
+    } else {
+      throw new IllegalStateException("Either readerFunction or batchReaderFunction must be set");
+    }
+  }
+
+  @FunctionalInterface
+  public interface ReaderFunction<D> {
```

Comment:
nit: this seems more like a `ReaderFactory` that can `create` a `ParquetValueReader `